test/rules/no-shadowed-variable/test.ts.lint

```diff
@@ -215,5 +216,95 @@ function creatorFunction(constructor: MyConstructor, filter: MyFunction): void {
 
 for (let i = 0; i < 3; i++) {
     let i = 34;
-        ~       [Shadowed variable: 'i']
+        ~       [err % ('i')]
+}
+
+class C<C> {
+        ~ [err % ('C')]
+    C<C>(C: C): C {
+      ~ [err % ('C')]
+         ~ [err % ('C')]
+        type C = C;
+             ~ [err % ('C')]
+        namespace C {
+                  ~ [err % ('C')]
+            interface C<C> {}
+                      ~ [err % ('C')]
+                        ~ [err % ('C')]
+        }
+    }
+}
+
+// allow nested namespaces
+namespace ns.C {}
+namespace ns {
+    export namespace C {}
+                     ~ [err % ('C')]
+}
+namespace ns {
+    namespace C {}
+              ~ [err % ('C')]
+}
+
+type Mapped<K, T> = {
+    [K in T]: K
+     ~ [err % ('K')]
+}
+
+// Allow parameters in signatures
+{
+    let param;
+    class TestClass {
+        myMethod(param): void;
+        myMethod(): void;
+        myMethod(param): void {}
+                 ~~~~~ [err % ('param')]
+    }
+    interface TestInterface {
+        (param): void;
+        myMethod(param): void
+        myMethod<param>(): void;
+                 ~~~~~ [err % ('param')]
+    }
 }
+
+// Handle variables in multiple child scopes
+{
+    {
+        {
+            let test;
+                ~~~~ [err % ('test')]
+        }
+        let test;
+            ~~~~ [err % ('test')]
+        {
+            let test;
+                ~~~~ [err % ('test')]
+        }
+    }
+    {
+        let test;
+            ~~~~ [err % ('test')]
+    }
+    let test;
+}
+
+// Allow mergeable declarations.
+interface I {}
+interface I {}
+
+// Ignore non-mergeable declarations as the compiler will complain about them anyway
```

Comment:
I don't see a compiler error (excepting the unused variable and implicit any) when I use that code. It's perfectly valid to write `function f() { const x = 0; return x; }`; the function `f` is shadowed.